### PR TITLE
[WIP] ipy3.1.0 update

### PR DESCRIPTION
**Pull request by cekees**
@ellisonbg and @jdfreder, I think I've reached (or exceeded) the current limits of my js ability. During the workhshop we started trying to update the roams notebook applications to ipython 3.1.0. I think there's some stuff in the js that needs more than simple  translation, which I'll  point out in an inline comment.

### CODE REVIEW

File: leafletwidget/leafletwidget.js [hunk 1]
```diff
@@ -351,45 +346,55 @@ require(["widgets/js/widget", "leaflet", "leaflet_draw"], function(WidgetManager
...
-            LeafletMapView.__super__.initialize.apply(this, arguments);
-            this.rendered = false;
+        initialize: function() {
+            console.log('MapView initialized');
+            LeafletMapView.__super__.initialize.apply(this,arguments);
+            this.rendered=false;
         },
     
         // Layer management
         update_layers: function (old_list, new_list) {
```
Comment (cekees) [inline leafletwidget/leafletwidget.js]: Ok, here's where I'm struggling. I saw a conversation about deprecating `do_diff` and adding `promises`, but I'm not sure how to update this critical bit of code.

File: leafletwidget/leafletwidget.js [hunk 2]
```diff
@@ -351,45 +346,55 @@ require(["widgets/js/widget", "leaflet", "leaflet_draw"], function(WidgetManager
...
+	    console.log('inside updating layers');
+	    console.log(old_list);
+	    console.log(new_list);
+            //_.difference(old_list,new_list).map(function(c) {this.remove_layer_model(c);});
+	    console.log('the difference');
+	    console.log(_.difference(new_list,old_list));
+	    console.log('adding new layers');
+            //_.difference(new_list,old_list).map(function(c) {$el.addClass(c);});
+	    this.add_layer_model(new_list[0]);
         },
```
Comment (cekees) [inline leafletwidget/leafletwidget.js]: Ok, here's where I'm struggling. I saw a conversation about deprecating do_diff and adding promises, but I'm not sure how to update this critical bit of code.

Comment (jdfreder) [inline leafletwidget/leafletwidget.js]: Ah, the layers should be managed by a view list.  https://github.com/ipython/ipywidgets/blob/master/ipython_widgets/static/widgets/js/widget.js#L716-L792

I can try to take a look at this tomorrow.

Comment (jdfreder) [inline leafletwidget/leafletwidget.js]: Used here https://github.com/ipython/ipywidgets/blob/master/ipython_widgets/static/widgets/js/widget_box.js#L25-L28

**Comment by cekees** [comment]
@jdfreder , thanks for the viewlist stuff. I'm working on the problem with the aligment of the map tiles now. I verified that the problem is not due to my upgrade of leaflet itself to the latest release in the sense that I get the same behavior with leaflet  0.7.2 and 0.7.3.

**Comment by cekees** [comment]
@ellisonbg and @jdfreder this is ready for review if you want  make this the current master. I'm going to make it master on the erdc-cm repo and tag it to ipython 3.1.0 somehow.

**Comment by jdfreder** [comment]
@ellisonbg I can't merge this so you'll have to (it's under your account).

**Pull request closed without merging**